File: BehavioralPatternTask/src/main/java/vendingMachineTaskJava/BuyCoffeeButton.java
package vendingMachineTaskJava;

public class BuyCoffeeButton {

    private CoinState coins;
    private int price = 2;

    public BuyCoffeeButton(CoinState state) {
        coins = state;
    }

    public void buy() {
        if (coins.hasEnoughCoins(price))
            System.out.println("Preparing coffee...");
        else
            System.out.println("Not enough to buy coffee...");
        coins.notify(this);
    }
}

File: BehavioralPatternTask/src/main/kotlin/vendingMachineTask/BuyCoffeeButton.kt
package vendingMachineTask

class BuyCoffeeButton(private val coins: CoinsState, val price: Int = 2) {

    fun buy() {
        if (coins.hasEnoughCoins(price))
            println("Preparing coffee...")
        else
            println("Not enough to buy coffee...")
    }
}